pkg/kubelet/dockertools/manager.go

```diff
@@ -1212,32 +1215,43 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
 		glog.Errorf("Failed to create symbolic link to the log file of pod %q container %q: %v", podFullName, container.Name, err)
 	}
 
-	// Set OOM score of POD container to lower than those of the other containers
-	// which have OOM score 0 by default in the pod. This ensures that it is
-	// killed only as a last resort.
-	containerInfo, err := dm.client.InspectContainer(string(id))
+	// Set OOM score of the container based on the priority of the container.
+	// Some containers are part of top-tier pods, and some containers are part of best-effort pods.
+	// Processes in best-effort pods should be killed first if the system runs out of memory.
+	// The main pod infrastructure container should be the last to get killed.
+	var oomScoreAdj int
+	if container.Name == PodInfraContainerName {
+		oomScoreAdj = qos.PodInfraOomAdj
+	} else {
+		oomScoreAdj = qos.GetPodOomAdjust(&pod.Spec)
+	}
+	fsManager := fs.Manager{
+		Cgroups: &configs.Cgroup{
```

Comment:
Ah you're right about the race. The current code is wrong for almost all setups though (systemd systems and when we put the containers at root which we do today). Let's try a third approach:

- Get the init PID from Docker (the existing code does this)
- Parse `/proc/<init PID>/cgroup` and find the path for `devices`
- Use that path as the container name

`devices` is the only cgroup Docker guarantees to exist and `/proc/<pid>/cgroup` gives us the current container of that cgroup hierarchy.